src/main/scala/viper/gobra/translator/Names.scala

```diff
@@ -6,17 +6,16 @@
 
 package viper.gobra.translator
 
+import viper.gobra.ast.{internal => in}
+import viper.gobra.theory.Addressability
+import viper.gobra.translator.interfaces.Context
+import viper.gobra.util.Violation
 import viper.silver.{ast => vpr}
 
 object Names {
   def returnLabel: String = "returnLabel"
 
-  private var freshCounter = 0
-  def freshName: String = {
-    val str = s"fn$$$$$freshCounter"
-    freshCounter += 1
-    str
-  }
+  def freshName(ctx: Context): String = s"fn$$$$${ctx.getAndIncrementFreshVariableCounter}"
```

Comment:
I like the idea of having an infinite iterator. I'm however not sure how exactly you want to separate the state because the iterator's current value depends on the context or rather the iterator's initial value should be settable by the context

Reply:
@jcp19 could you have a look at `Context.scala` and `ContextImpl.scala` and let me know whether you like the solution?

Reply:
Fine by me